### ISSUE TITLE
Model.find() and Model.update()

### ISSUE THREAD

**Issue by molexx** [iliakan/cls-mongoose#3]
Should Model.find() and Model.update() work?

On mongoose 3.8 and 3.9, Model.find() works OK but in Model.update() the context does not seem to be carried over to the callback.
On mongoose 4.0 and 4.1 Model.find() does not work either.

I've added an example in a test to demonstrate:

https://github.com/molexx/cls-mongoose/blob/master/test/mongoose-cls-test.js

Note if you run this it will try to connect to 'mongodb://localhost/mongoose-cls-test'.

**Comment by iliakan** [comment]
Yes, should work, I can look into it, or...

The latest mongoose supports native promises as well as other promise libraries. Maybe that can help?

**Comment by molexx** [comment]
We're using 3.8 on a large project that has been in production for a while and aren't keen to upgrade as mongoose 4 has breaking changes.

If you could take a look please that would be great. I've been experimenting - see gist - it seems to work but I'm not really happy about it because your promise wrapper ought to cover all scenarios and my wrapper needs to be individually added to every function in the Model. I've since found that distinct() has the same problem, I expect others will too.

https://gist.github.com/molexx/2f6b3b7d132b9fbf23cd

On the other hand if you think that's the correct way to do it I'm happy to formalise and send a pull request.

Thanks.

**Comment by iliakan** [comment]
You said it doesn't work for mongoose 4.0 too?

**Comment by molexx** [comment]
Yes, actually it's worse in 4.0 - Model.find() works in 3.8 but not 4.0.

**Comment by iliakan** [comment]
The internal mechanics has changed, please retest the code against the modern mongoose.

I had to make it compatible with 4, guess the mechanics is different in 4 vs 3, so the patch is different.

**Comment by molexx** [comment]
Thanks for this. The tests now pass for both 3.8 and 4.1! Without my code, so your new code has fixed 3.8 too.

But Model.aggregate() and distinct() don't work.
If I add back in your original 'promise.on()' shim then aggregate() does work in 3.8 but doesn't work in 4.1.

I've updated my fork from yours then added promise.on() back in and added tests for aggregate() and distinct(): https://github.com/molexx/cls-mongoose

**Comment by iliakan** [comment]
Great you found it out. Surely must be fixed.

Let's get a feedback from mongoose https://github.com/Automattic/mongoose/issues/3313

Maybe CLS-integration can be somehow aknowledged from the core.

**Comment by molexx** [comment]
In my fork it seems to be fixed for 3.8 but not 4.1.

**Comment by iliakan** [comment]
Meanwhile I expanded tests. So far all pass for me, mongoose@4:

![](http://ilyakantor.ru/screen/2015-08-28_1345.png)

**Comment by molexx** [comment]
Weird, my test is still failing on aggregate() and count() with mongoose 4.1.3.

Can you try running my test.js please? I've updated it to be separate tests.

https://github.com/molexx/cls-mongoose/blob/master/test.js

**Comment by iliakan** [comment]
Your test fails indeed. I added it to mocha and it passes.

Would be great if you could modify my tests, to make them fail (pushed more).

**Comment by molexx** [comment]
I had to do some faffing to get the mocha tests to run - as they use generators I had to upgrade node (we are using 0.10) and I had to add --harmony to mocha.opts.

The mocha tests all pass, but if I edit index.js and comment out both shim calls - so cls-mongoose isn't doing anything - the tests still all pass which they definitely shouldn't. So I don't think the mocha tests are doing any testing.
I prefer the tap tests as other cls- modules seem to use tap and they work on node 0.10 too :-)

**Comment by iliakan** [comment]
I pushed simple tests for `Model#aggregate` with mocha.
Without `yield`.

They pass without CLS :/
Please try, first with the modern node.

From my experience, by the way, I met lots of "incorrect CLS values" with unpatched mongoose & parallel requests. But never a "lost CLS value".

**Comment by molexx** [comment]
It might be something to do with the context being entered in mocha's before() which surrounds all the tests? Not sure, but I've 'fixed' it in that it agrees with the tap tests:

https://github.com/molexx/cls-mongoose/blob/master/test/index.js

My node also didn't like the 'const' at the top of the file which is ok if I make it a 'var'.

If you comment out line 9 and so remove cls-mongoose then only update() works but find() and aggregate() don't work, with line 9 uncommented find() and update() both work but aggregate() still does not work.
